### DISCUSSION TITLE
手把手教你完成 task 9 （拿 10U 奖励）

### DISCUSSION THREAD

**Discussion by linghuccc** [openbuildxyz/Web3-Frontend-Bootcamp#1198] [Materials]
虽然这次的 task 很多，而且有一定难度，但是最后几个 task，特别是 **task 7, 8, 9**，可以说是给大家送**福利**来的，希望大家**优先**去做，千万不要放过了！

**准备工作：**
在做 task 9 之前，请大家先准备好一个钱包账户，一个电子邮件信箱（用来接收验证码），一个已经部署好的合约（需要合约地址）。

**第一步：创建 graph 账户**
进入下面的网页：
https://thegraph.com/studio/ ，点击 “Connect Wallet” （连接钱包）
![g1](https://github.com/user-attachments/assets/a1ec09dd-54bf-4307-b137-36120f57428e)

连接钱包之后，需要签名
![g2](https://github.com/user-attachments/assets/9cdeb411-3e12-477d-9d69-1ead57793101)

签名之后，需要填写电子邮件地址，如何点击“Send verification code” （发送验证码）
![g3](https://github.com/user-attachments/assets/f07cfefc-e92a-4645-af3a-3c68272ef7d4)

去电子邮件信息，找到验证码，并填入，然后点击“Verify” （验证）
![g4](https://github.com/user-attachments/assets/64ddd0ed-a8a5-47d0-8cb3-28f7280bf259)

**第二步：创建在线 Subgraph**
点击网页右边的“Create a Subgraph” （创建一个 Subgraph）
![g5](https://github.com/user-attachments/assets/057b3474-6e13-4321-a782-550fa69668d0)

输入 Subgraph 的名字，然后点击 “Create Subgraph” （创建 Subgraph）
![g6](https://github.com/user-attachments/assets/40155226-cab3-4eda-9220-e335498b76b8)

生成之后，你会发现 “Save” （保存）按钮是灰的，必须在 “Subgraph Description” （Subgraph 描述）部分填写内容，  “Save” （保存）按钮才会被 enable。
![g7](https://github.com/user-attachments/assets/7929d551-5127-46e5-999f-394066402876)

点击 "Save" 保存了 Subgraph 之后，就可以根据右边的 commands 在自己电脑上创建 Subgraph 了。
![g8](https://github.com/user-attachments/assets/c38c956a-f7e1-407d-9ed3-3db98e4fba4f)

**第三步：创建本地 Subgraph**
首先是安装 graph，虽然网页上有 npm 和 yarn 2 个命令，但我建议用 npm 命令去安装（我本人在用 yarn 安装的时候，遇到一点问题，npm 则 OK）。
![g9](https://github.com/user-attachments/assets/030bd133-98da-4618-9dba-9e948f48badc)

然后 init 本地的 Subgraph：
![g10](https://github.com/user-attachments/assets/9dbc8139-d68a-471f-8690-4e34625ed483)

当问到 Protocol 的时候，选择 ethereum：
![g11](https://github.com/user-attachments/assets/296ebb42-607d-495d-8d70-040aab29d00c)

Subgraph slug，则填入上一步的 Subgraph slug：
![g12](https://github.com/user-attachments/assets/75410156-2b45-4cfb-baf9-2dc91fa38eaa)
![g13](https://github.com/user-attachments/assets/5fdfda47-3de8-4c9c-a3ae-7df2dd8b503a)

Directory to create the subgraph in，则填入你希望生成的本地目录的文件夹名：
![g14](https://github.com/user-attachments/assets/3aa7a1e8-bd4c-4036-b03e-85afa8249224)

选择网络时，用上下箭头选中已部署合约的网络：
![g15](https://github.com/user-attachments/assets/6c7ec4ea-5fb8-4ad9-89bb-3051222fcb30)

Contract address，则填入已部署的合约地址：
![g16](https://github.com/user-attachments/assets/1dce7582-22ae-4b7e-bebb-7ac48e8cb546)

由于网络的问题，在 Fetching Start Block 和 Fetching Contract Name 的时候，可能经常会报错，但是不要紧，每次直接按回车，让他 retry 就好了：
![g17](https://github.com/user-attachments/assets/bbca6332-95e5-4c97-baff-d9f23d2b13b0)

拿到 Start Block 和 Contract Name 之后，Index contract events as entities 不用改，直接回车就可以，之后等待一段时间，会问 Add another contract?，点击 “n”，Subgraph 就创建完成了。
![g18](https://github.com/user-attachments/assets/7b9eee63-ea69-4a1f-980e-854ca82d5b37)

**第四步：Deploy Subgraph**
本地生成的 Subgraph 文件夹，里面最主要的是三个文件：src/ 里面的 .ts 文件，schema.graphql 文件和 subgraph.yaml 文件，有兴趣的可以打开看一下：
![g19](https://github.com/user-attachments/assets/92db10a7-f5e2-4db4-ae4f-6d69def90ea1)

Authenticate within the CLI：
![g20](https://github.com/user-attachments/assets/eba557d0-bf03-42a1-b4e5-8e141f407fb1)

Build Subgraph：
![g21](https://github.com/user-attachments/assets/e6fbbb48-4980-41fb-b0d1-1c8f1d6c3880)

Deploy Subgraph：
![g22](https://github.com/user-attachments/assets/1551696b-4ee8-4fbd-97d1-2a299c22f68e)

在 Deploy 的时候，会问版本号，填你想用的版本号就可以：
![g23](https://github.com/user-attachments/assets/5b6c2d23-4954-4f70-8911-467b5b164182)

**第五步：Play around**
Deploy 成功之后，刷新页面，会发现在线的 Subgraph 页面有一些变化：
![g24](https://github.com/user-attachments/assets/1904657d-a6a2-4841-975a-d2a18d1c9a7d)

这时点击 Logs，会发现多了很多东西：
![g25](https://github.com/user-attachments/assets/8673029f-36f0-4541-8243-ccf76076fc57)

点击 Playground，通过修改左边的代码，然后点击中间的按钮，右边就会显示出我们想要知道的信息：
![g26](https://github.com/user-attachments/assets/b61ab276-1732-43ad-809d-038555414571)

还在等待什么，快点截图并提交 PR 吧~

**Comment by linghuccc** [comment]
**提交 PR 时要注意**：

部署的子图的链接，应该填写包含 api 的 URL（任一即可），而不是你当前网页的 URL。
![g27](https://github.com/user-attachments/assets/d88a6063-0fe9-470c-a308-13a32f0f857d)

**Comment by zhangloveyan** [comment]
Tips：
合约没有验证时，需要本地加载 abi.json 文件
window 格式：‪D:\\zz\\abi.json
要手动加 \ ，否则报错识别不到。

**Comment by linghuccc** [comment]
在运行页面命令时，请如下图所示，尽量使用拷贝按钮（特别是第 4 步）
![g29](https://github.com/user-attachments/assets/b0e145a1-5ae9-40d7-82dd-e8f037dc2090)

**Comment by posuihushui** [comment]
网络有问题可以参考这个，https://github.com/smartcontractkit/full-blockchain-solidity-course-js/discussions/2247#discussioncomment-5496669

**Comment by qiaopengjun5162** [comment]
graph deploy --studio nftmarketsepolia

Error: Failed to upload file to IPFS: <!DOCTYPE html>

**Reply by linghuccc** [reply]
可能是网络问题，可以考虑在其他时间重试一下；
可以在 the graph 官方群里询问一下。

**Reply by qiaopengjun5162** [reply]
这个问题解决了， 本地起一个IPFS ， 执行 graph deploy --studio nftmarketsepolia --ipfs http://127.0.0.1:5001
